lib/warren/config/consumers.rb

```diff
@@ -86,6 +91,23 @@ def queue_config(queue_name, bindings, dead_letter_exchange = nil)
         }
       end
 
+      # rubocop:todo Metrics/MethodLength
+      def delay_exchange_configuration(ttl:, original_queue:, consumer_name:)
+        return {} if ttl.nil? || ttl.zero?
+
+        {
+          'exchange' => { 'name' => "#{consumer_name}.delay", 'options' => { type: 'fanout', durable: true } },
+          'bindings' => [{
+            'queue' => { 'name' => "#{consumer_name}.delay", 'options' => {
+              durable: true, arguments: {
+                'x-dead-letter-exchange' => '', 'x-message-ttl' => ttl, 'x-dead-letter-routing-key' => original_queue
+              }
+            } }, 'options' => {}
+          }]
+        }
+      end
```

Comment:
Bit lost on me where/how this bit gets used too. Looks good, just can't place it in my head with the exchange `"#{consumer_name}.delay"`

Reply:
It gets called above in add consumer, which is used by the `warren consumer add` CLI task if you say you want a delay queue. This populates the warren_consumers.yml file.

This populated the 'delay' section which configures a delay exchange. By default we label delay exchanges after the consumer that uses them. Hence "#{consumer_name}.delay", this then attaches to a queue of the same name. Its a fanout exchange, which means we don't need to specify a routing key.

The next important section is:
```
arguments: {
                'x-dead-letter-exchange' => '', 'x-message-ttl' => ttl, 'x-dead-letter-routing-key' => original_queue
```

This is the configuration used for the delay queue.

Here we set a time to live for all messages, and when they expire they get dead-lettered to the exchange specified in x-dead-letter-exchange.

If we sent them back the the original exchange, then we'd risk duplicate messages on all other queues. But fortunately RabbitMQ provides a nice solution.

The default exchange is used whenever the exchange name is blank. This is a special exchange, that is bound to all queues, using a routing key based on the queue name. So we just simple set our 'x-dead-letter-routing-key' to match the original queue, and the message will pop back on there.

I'll add a bit of documentation for this somewhere actually, as the default exchange magic took a while to work out, but saved me a massive headache.